src/SymbolTable.ts

```diff
@@ -0,0 +1,101 @@
+import type { Identifier } from './lexer/Token';
+import type { BscType } from './types/BscType';
+import { DynamicType } from './types/DynamicType';
+import { UninitializedType } from './types/UninitializedType';
+
+
+/**
+ * Stores the types associated with variables and functions in the Brighterscript code
+ * Can be part of a hierarchy, so lookups can reference parent scopes
+ */
+export class SymbolTable {
+
+    private symbols = new Map<string, BscSymbol[]>();
+
+    constructor(
+        private parent?: SymbolTable | undefined
+    ) { }
+
+    /**
+     * Sets the parent table for lookups
+     *
+     * @param [parent]
+     */
+    setParent(parent?: SymbolTable) {
+        this.parent = parent;
+    }
+
+    /**
+     * Gets the name/type pair for a given named variable or function name
+     * If the identifier is not in this table, it will check the parent
+     *
+     * @param  name the name to lookup
+     * @returns An array of BscSymbols - one for each time this symbol had a type implicitly defined
+     */
+    getSymbol(name: string): BscSymbol[] {
+        const key = name.toLowerCase();
+        return this.symbols.get(key) ?? this.parent?.getSymbol(key);
+    }
+
+    /**
+     * Adds a new symbol to the table
+     * @param name
+     * @param  type
+     */
+    addSymbol(name: Identifier, type: BscType) {
```

Comment:
Based on [this comment](https://github.com/rokucommunity/brighterscript/pull/380/files#r605556107), change this method signature to:
```
addSymbol(name: string, type: BscType, range?: Range) {
```